mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxUpdateParentAsLookupListener.java

```diff
@@ -0,0 +1,103 @@
+package org.apache.james.mailbox.store.event;
+
+import java.util.List;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.acl.ACLDiff;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.UnsupportedRightException;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.event.EventFactory.MailboxLookupRightUpdate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MailboxUpdateParentAsLookupListener implements MailboxListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MailboxUpdateParentAsLookupListener.class);
+
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+    private final StoreRightManager storeRightManager;
+
+    @Inject
+    public MailboxUpdateParentAsLookupListener(MailboxSessionMapperFactory mailboxSessionMapperFactory, StoreRightManager storeRightManager) {
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.storeRightManager = storeRightManager;
+    }
+
+    @Override
+    public ListenerType getType() {
+        return ListenerType.ONCE;
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return ExecutionMode.ASYNCHRONOUS;
+    }
+
+    @Override
+    public void event(Event event) {
+        if (event instanceof EventFactory.MailboxLookupRightUpdate) {
+            MailboxLookupRightUpdate aclUpdateEvent = (MailboxLookupRightUpdate) event;
+
+            MailboxSession mailboxSession = aclUpdateEvent.getSession();
+            MailboxPath mailboxPath = aclUpdateEvent.getMailboxPath();
+            List<MailboxPath> mailboxPaths = mailboxPath.getHierarchyLevels(mailboxSession.getPathDelimiter());
+            mailboxPaths.stream()
+                .filter(hierarchyMailboxPath -> !hierarchyMailboxPath.equals(mailboxPath))
+                .forEach(parentMailboxPath -> {
+                    updateLookupRight(parentMailboxPath, mailboxSession, aclUpdateEvent.getAclDiff());
+                });
+        }
+    }
+
+    private void updateLookupRight(MailboxPath mailboxPath, MailboxSession session, ACLDiff aclDiff) {
+        aclDiff.addedEntries()
+            .forEach(entry -> {
+                updateRight(mailboxPath, session, entry);
+        });
+        aclDiff.changedEntries()
+            .forEach(entry -> {
+                updateRight(mailboxPath, session, entry);
+        });
+    }
+
+    private void updateRight(MailboxPath mailboxPath, MailboxSession session, MailboxACL.Entry entry) {
+        MailboxACL.ACLCommand mailboxACLCommand = null;
+        try {
+            mailboxACLCommand = MailboxACL.command()
```

Comment:
Could we skip that update if the guy already has lookup right?

Reply:
Also you are granting rights **every time** the ACL changes. Even if the new version does not contains lookup right.